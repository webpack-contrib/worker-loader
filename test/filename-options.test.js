import path from 'path';

import { customAlphabet } from 'nanoid';

import {
  compile,
  getCompiler,
  getErrors,
  getModuleSource,
  getResultFromBrowser,
  getWarnings,
} from './helpers';

describe('"filename" option', () => {
  it('should work ("string")', async () => {
    const compiler = getCompiler('./chunks/entry.js', {
      filename: '[name].custom.worker.js',
    });
    const stats = await compile(compiler);
    const result = await getResultFromBrowser(stats);

    expect(getModuleSource('./chunks/worker.js', stats)).toMatchSnapshot(
      'module'
    );
    expect(result).toMatchSnapshot('result');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work ("function")', async () => {
    const compiler = getCompiler('./chunks/entry.js', {
      filename: () => '[name].custom.worker.js',
    });
    const stats = await compile(compiler);
    const result = await getResultFromBrowser(stats);

    expect(getModuleSource('./chunks/worker.js', stats)).toMatchSnapshot(
      'module'
    );
    expect(result).toMatchSnapshot('result');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work and respect the "output.filename" default value option', async () => {
    const nanoid = customAlphabet('1234567890abcdef', 10);
    const compiler = getCompiler(
      './chunks/entry.js',
      {},
      {
        output: {
          path: path.resolve(__dirname, './outputs', `test_${nanoid()}`),
          chunkFilename: '[name].chunk.js',
          publicPath: '',
        },
        module: {
          rules: [
            {
              test: /worker\.js$/i,
              rules: [
                {
                  loader: path.resolve(__dirname, '../src'),
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);
    const result = await getResultFromBrowser(stats);

    expect(getModuleSource('./chunks/worker.js', stats)).toMatchSnapshot(
      'module'
    );
    expect(result).toMatchSnapshot('result');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work and respect the "output.filename" option ("string")', async () => {
    const nanoid = customAlphabet('1234567890abcdef', 10);
    const compiler = getCompiler(
      './chunks/entry.js',
      {},
      {
        output: {
          path: path.resolve(__dirname, './outputs', `test_${nanoid()}`),
          filename: '[name].custom.js',
          chunkFilename: '[name].chunk.js',
          publicPath: '',
        },
        module: {
          rules: [
            {
              test: /worker\.js$/i,
              rules: [
                {
                  loader: path.resolve(__dirname, '../src'),
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);
    const result = await getResultFromBrowser(stats);

    expect(getModuleSource('./chunks/worker.js', stats)).toMatchSnapshot(
      'module'
    );
    expect(result).toMatchSnapshot('result');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work and respect the "output.filename" option ("function")', async () => {
    const nanoid = customAlphabet('1234567890abcdef', 10);
    const compiler = getCompiler(
      './chunks/entry.js',
      {},
      {
        output: {
          path: path.resolve(__dirname, './outputs', `test_${nanoid()}`),
          filename: (pathData) => {
            if (/worker\.js$/.test(pathData.chunk.entryModule.resource)) {
              return '[name].custom.worker.js';
            }

            return '[name].js';
          },
          publicPath: '',
        },
        module: {
          rules: [
            {
              test: /worker\.js$/i,
              rules: [
                {
                  loader: path.resolve(__dirname, '../src'),
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);
    const result = await getResultFromBrowser(stats);

    expect(getModuleSource('./chunks/worker.js', stats)).toMatchSnapshot(
      'module'
    );
    expect(result).toMatchSnapshot('result');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });
});
