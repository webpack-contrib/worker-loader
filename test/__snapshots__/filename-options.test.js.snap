// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"filename" option should work ("function"): errors 1`] = `Array []`;

exports[`"filename" option should work ("function"): module 1`] = `
"export default function Worker_fn() {
  return new Worker(__webpack_public_path__ + \\"worker.custom.worker.js\\");
}
"
`;

exports[`"filename" option should work ("function"): result 1`] = `"{\\"postMessage\\":true,\\"onmessage\\":true}"`;

exports[`"filename" option should work ("function"): warnings 1`] = `Array []`;

exports[`"filename" option should work ("string"): errors 1`] = `Array []`;

exports[`"filename" option should work ("string"): module 1`] = `
"export default function Worker_fn() {
  return new Worker(__webpack_public_path__ + \\"worker.custom.worker.js\\");
}
"
`;

exports[`"filename" option should work ("string"): result 1`] = `"{\\"postMessage\\":true,\\"onmessage\\":true}"`;

exports[`"filename" option should work ("string"): warnings 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" default value option: errors 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" default value option: module 1`] = `
"export default function Worker_fn() {
  return new Worker(__webpack_public_path__ + \\"worker.worker.js\\");
}
"
`;

exports[`"filename" option should work and respect the "output.filename" default value option: result 1`] = `"{\\"postMessage\\":true,\\"onmessage\\":true}"`;

exports[`"filename" option should work and respect the "output.filename" default value option: warnings 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" option ("function"): errors 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" option ("function"): module 1`] = `
"export default function Worker_fn() {
  return new Worker(__webpack_public_path__ + \\"worker.custom.worker.js\\");
}
"
`;

exports[`"filename" option should work and respect the "output.filename" option ("function"): result 1`] = `"{\\"postMessage\\":true,\\"onmessage\\":true}"`;

exports[`"filename" option should work and respect the "output.filename" option ("function"): warnings 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" option ("string"): errors 1`] = `Array []`;

exports[`"filename" option should work and respect the "output.filename" option ("string"): module 1`] = `
"export default function Worker_fn() {
  return new Worker(__webpack_public_path__ + \\"worker.custom.worker.js\\");
}
"
`;

exports[`"filename" option should work and respect the "output.filename" option ("string"): result 1`] = `"{\\"postMessage\\":true,\\"onmessage\\":true}"`;

exports[`"filename" option should work and respect the "output.filename" option ("string"): warnings 1`] = `Array []`;
